Show GA4GH WDL import URL for use with subworkflows; distinguish public methods [GAWB-2919, GAWB-2543]

aednichols opened the pull request:
I wrote some ClojureScript to automatically generate GA4GH URLs for methods because I got tired of doing it manually. Let's put it in the UI to make users' lives easier.

~GAWB-2814~ GAWB-2919: Received PO approval that this is a thing we should do, and UX approval to do it in this general way.

GAWB-2543: Just something I whipped up since I was already in the right place. Needs UX review.

Looks like this:

![screenshot-2017-11-17 firecloud broad institute](https://user-images.githubusercontent.com/1087943/32972895-9aacdab6-cbc2-11e7-93c4-4a693f8c2e49.png)

Related PRs: https://github.com/broadinstitute/agora/pull/226, https://github.com/broadinstitute/firecloud-orchestration/pull/546

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

aednichols committed:
Show GA4GH WDL import URL for use with subworkflows

aednichols commented:
Note to self: hide if not public

davidangb commented:
As you said, we should definitely only show this for public methods/snapshots, as the ga4gh url will fail for private methods.

Simultaneously, I think we should expand this a bit and show a "public" badge for those public methods. Can we combine this PR with GAWB-2543?

aednichols committed:
Only show GA4GH link for public methods

aednichols committed:
Rudimentary impl of GAWB-2543

aednichols commented:
Note: I added the permissions AJAX before reading David's note about adding public to the get-method API. Open to changing.

I can also do some more work to deduplicate the perms call I added here and the perms call in the perms modal.

davidangb commented on src/cljs/main/broadfcui/page/method_repo/method/details.cljs:
```diff
@@ -197,7 +217,14 @@
+                   (fn [{:keys [success? parsed-response]}]
+                     (if success?
+                       (swap! state assoc :is-public (= (:role (first (filter #(= "public" (:user %)) parsed-response))) "READER"))
+                       (swap! state assoc :is-public false))))})))})
```
I do think this needs to move into the back end, with agora returning a "public:true" flag as appropriate. Most importantly - this ajax call will only succeed if you are owner on the snapshot. If you only have reader - and for public methods, most of the world will only be readers - this results in a 403, and we won't display the public icon.

zarsky-broad replied:
I agree with David- this shouldn't be a frontend check...

zarsky-broad commented on src/cljs/main/broadfcui/page/method_repo/method/details.cljs:
```diff
@@ -18,6 +20,7 @@
    [broadfcui.page.workspace.method-configs.synchronize :as mc-sync]
    [broadfcui.page.workspace.workspace-common :as ws-common]
    [broadfcui.utils :as utils]
+   [broadfcui.common.icons :as icons]
```
nit: order

zarsky-broad commented on src/cljs/main/broadfcui/page/method_repo/method/details.cljs:
```diff
@@ -68,15 +71,19 @@
           (this :-render-snapshot-selector)]
+         (when (:is-public @state)
+           [:span {:style {:alignSelf "center" :marginLeft "2.5rem"}}
+            (icons/render-icon {} :public)
```
put a bit of a margin-right on the icon, rather than starting the text with a space.

aednichols replied:
Thanks, the space felt kinda dumb

zarsky-broad commented on src/cljs/main/broadfcui/page/method_repo/method/details.cljs:
```diff
@@ -105,8 +112,21 @@
                WDL (react/create-element
-                    [WDLViewer
-                     {:ref WDL :wdl (:payload selected-snapshot)}])
+                    [:div {:style {}}
```
don't need an empty `:style`

zarsky-broad commented on src/cljs/main/broadfcui/page/method_repo/method/details.cljs:
```diff
@@ -105,8 +112,21 @@
+                    [:div {:style {}}
+                     [WDLViewer
+                      {:ref WDL :wdl (:payload selected-snapshot)}]
+                     (if (:is-public @state)
```
`when`

zarsky-broad commented on src/cljs/main/broadfcui/page/method_repo/method/details.cljs:
```diff
@@ -105,8 +112,21 @@
+                        "Import URL for this WDL: "
+                        (let [link (str (config/api-url-root)
+                                        "/ga4gh/v1/tools/"
+                                        (:namespace selected-snapshot)
```
`:keys` these out

zarsky-broad commented on src/cljs/main/broadfcui/page/method_repo/method/wdl.cljs:
```diff
@@ -8,7 +8,7 @@
   {:render
    (fn [{:keys [props]}]
-     [:div {:style {:margin "2.5rem 1.5rem"}}
+     [:div {:style {:marginTop "2.5rem" :marginLeft "1.5rem" :marginRight "1.5rem" :marginBottom "1.0rem"}}
```
just `1 rem`

aednichols replied:
`{:margin "1.0rem"}`? That does change the horizontal position of CodeMirror on the page (in a way that looks better to me)

aednichols replied:
(Though the URL label is no longer aligned)

zarsky-broad replied:
Ah, I just meant that you didn't need to specify "1.0rem" for the `marginBottom`

aednichols replied:
Oooooh, that was a woosh on my part. Thanks

aednichols committed:
Eliminate unnecessary AJAX call

aednichols committed:
CLJS style

aednichols commented:
jenkins retest

aednichols committed:
CLJS style

aednichols committed:
Pad the right way

aednichols committed:
Isaac concision improvement

aednichols commented:
jenkins retest

davidangb reviewed (commented):
:+1: for the code/syntax.

As for the UX - I bet you've gotten this reviewed already, but: I often see links like this held inside a disabled `[:input {}]` field, so end users can easily triple-click to select the whole thing, then copy it for pasting elsewhere. I think we'd still want the icon to open the link in another tab, though. 

So maybe something like this?
```
[:input {:type "text" :disabled true :value link }]
(links/create-external {:href link} "")
```
note: I didn't actually *try* the above code so it might not work ... 

_EDIT: see, for instance, stackoverflow ... click the "share" link under a question or answer._

zarsky-broad commented:
Good call, I specifically like how GitHub does it in the clone repo popover.

aednichols committed:
Isaac functional improvement

aednichols committed:
Isaac: "now it looks like a thing"

davidangb reviewed (approved):
:+1:

aednichols commented:
jenkins retest
